File: sdk/golang/example/example_ws.go
package main

import (
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/api"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/futures/futurespublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/spot/spotpublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"os"
	"time"
)

func WsExample() {
	// Use the default logger or supply your custom logger
	defaultLogger := logger.NewDefaultLogger()
	logger.SetLogger(defaultLogger)

	// Retrieve API secret information from environment variables
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	// Set specific options, others will fall back to default values
	wsOption := types.NewWebSocketClientOptionBuilder().Build()

	// Create a client using the specified options
	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithWebSocketClientOption(wsOption).
		Build()
	client := api.NewClient(option)

	// Get the websocket service
	wsService := client.WsService()

	spotWsExample(wsService.NewSpotPublicWS())
	futuresWsExample(wsService.NewFuturesPublicWS())
}

func spotWsExample(spotPublicWs spotpublic.SpotPublicWS) {
	err := spotPublicWs.Start()
	if err != nil {
		logger.GetLogger().Error("failed to start spot public websocket: %v", err)
		return
	}
	defer spotPublicWs.Stop()

	tickerEventCallback := func(topic string, subject string, data *spotpublic.TickerEvent) error {
		// Process logic
		logger.GetLogger().Infof("received ticker event %s %s %s %s %d %s",
			topic, subject, data.Sequence, data.Price, data.Time, data.Size)
		return nil
	}

	// Get ticker
	subId, err := spotPublicWs.Ticker([]string{"BTC-USDT"}, tickerEventCallback)

	if err != nil {
		logger.GetLogger().Fatalf("subscribe error: %v", err)
		return
	}

	// Triggered when certain conditions are met
	<-time.After(time.Second * 10)

	// Unsubscribe by sub id
	err = spotPublicWs.UnSubscribe(subId)
	if err != nil {
		logger.GetLogger().Fatalf("unsubscribe error: %v, id: %s", err, subId)
	}
}

func futuresWsExample(futuresPublicWs futurespublic.FuturesPublicWS) {
	err := futuresPublicWs.Start()
	if err != nil {
		logger.GetLogger().Fatalf("failed to start futures public websocket: %v", err)
		return
	}

	defer futuresPublicWs.Stop()

	tickerEventV2Callback := func(topic string, subject string, data *futurespublic.TickerV2Event) error {
		logger.GetLogger().Infof("received ticker event %+v", data.ToMap())
		return nil
	}

	//  Get Ticker
	subId, err := futuresPublicWs.TickerV2("XBTUSDTM", tickerEventV2Callback)

	if err != nil {
		logger.GetLogger().Fatalf("subscribe error: %v", err)
		return
	}
	// Triggered when certain conditions are met
	<-time.After(time.Second * 10)
	err = futuresPublicWs.UnSubscribe(subId)
	if err != nil {
		logger.GetLogger().Fatalf("unsubscribe error: %v, id: %s", err, subId)
	}
}

File: sdk/python/example/example_ws.py
import logging
import os
import time

from kucoin_universal_sdk.api.client import DefaultClient
from kucoin_universal_sdk.generate.futures.futures_public.model_ticker_v2_event import TickerV2Event
from kucoin_universal_sdk.generate.futures.futures_public.ws_futures_public import FuturesPublicWS
from kucoin_universal_sdk.generate.spot.spot_public.model_ticker_event import TickerEvent
from kucoin_universal_sdk.generate.spot.spot_public.ws_spot_public import SpotPublicWS
from kucoin_universal_sdk.model.client_option import ClientOptionBuilder
from kucoin_universal_sdk.model.constants import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model.websocket_option import WebSocketClientOptionBuilder


def ws_example():
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )

    # Retrieve API secret information from environment variables
    key = os.getenv("API_KEY", "")
    secret = os.getenv("API_SECRET", "")
    passphrase = os.getenv("API_PASSPHRASE", "")

    # Set specific options, others will fall back to default values
    ws_client_option = (
        WebSocketClientOptionBuilder().build()
    )

    # Create a client using the specified options
    client_option = (
        ClientOptionBuilder()
        .set_key(key)
        .set_secret(secret)
        .set_passphrase(passphrase)
        .set_websocket_client_option(ws_client_option)
        .set_spot_endpoint(GLOBAL_API_ENDPOINT)
        .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
        .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
        .build()
    )
    client = DefaultClient(client_option)
    # Get the websocket service
    kucoin_ws_service = client.ws_service()
    spot_ws_example(kucoin_ws_service.new_spot_public_ws())
    futures_ws_example(kucoin_ws_service.new_futures_public_ws())


def spot_ws_example(spot_public_ws: SpotPublicWS):
    try:
        # Start WebSocket
        spot_public_ws.start()
    except Exception as e:
        logging.error(f"failed to start spot public websocket: {e}")
        return

    try:
        def ticker_event_callback(topic: str, subject: str, data: TickerEvent) -> None:
            # Process logic
            logging.info(
                f"received ticker event {topic} {subject} {data.sequence} {data.price} {data.time} {data.size}")

        # Get ticker
        try:
            sub_id = spot_public_ws.ticker(["BTC-USDT"], ticker_event_callback)
        except Exception as e:
            logging.fatal(f"subscribe error: {e}")
            return

        # Triggered when certain conditions are met
        time.sleep(10)

        # Unsubscribe by sub id
        try:
            spot_public_ws.unsubscribe(sub_id)
        except Exception as e:
            logging.fatal(f"unsubscribe error: {e}, id: {sub_id}")
    finally:
        spot_public_ws.stop()


def futures_ws_example(futures_public_ws: FuturesPublicWS):
    try:
        # Start WebSocket
        futures_public_ws.start()
    except Exception as e:
        logging.fatal(f"failed to start futures public websocket: {e}")
        return

    try:
        def ticker_event_v2_callback(topic: str, subject: str, data: TickerV2Event) -> None:
            logging.info(f"received ticker event {data.to_json()}")

        # Get Ticker
        try:
            sub_id = futures_public_ws.ticker_v2("XBTUSDTM", ticker_event_v2_callback)
        except Exception as e:
            logging.fatal(f"subscribe error: {e}")
            return
        # Triggered when certain conditions are met
        time.sleep(10)

        # Unsubscribe
        try:
            futures_public_ws.unsubscribe(sub_id)
        except Exception as e:
            logging.fatal(f"unsubscribe error: {e}, id: {sub_id}")
    finally:
        futures_public_ws.stop()


if __name__ == "__main__":
    ws_example()
